### PR TITLE
Filtered operators which have no outputs while getting schema and added test.

### LINKED ISSUE #2156: Loading of same model for tensorflow throw exception
just run in two different threads:
`var loadModelSchema = TensorFlowUtils.GetModelSchema(mlContext, modelLocation);`
it will throw exception.

### PR DESCRIPTION
**Pull request by zeahmed**
This PR fixes #2156.

Please see the issue for more details and the solution of the bug.

### CODE REVIEW

File: src/Microsoft.ML.TensorFlow/TensorFlow/TensorflowUtils.cs [hunk 1]
```diff
@@ -40,7 +40,10 @@ internal static Schema GetModelSchema(IExceptionContext ectx, TFGraph graph, str
 
                 // If the type is not supported in ML.NET then we cannot represent it as a column in an Schema.
                 // We also cannot output it with a TensorFlowTransform, so we skip it.
-                if (mlType == null)
+                // Furthermore, 'NoOp' operator usually has 'Unknown' type.
+                // But sometimes it can have valid type upon querying model for multiple times such as 'Float'.
+                // This behavior is inconsistent in TensorFlow runtime (#2156). Therefore, filtering all 'NoOp' operators in the model.
```
Comment (eerhardt) [inline src/Microsoft.ML.TensorFlow/TensorFlow/TensorflowUtils.cs]: (nit) - can you give a full URL for `(#2156)`? That way we know exactly which to look up.

File: test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs [hunk 2]
```diff
@@ -797,6 +798,27 @@ public void TensorFlowTransformCifarSavedModel()
...
+            var t3 = new Thread(() => TensorFlowUtils.GetModelSchema(mlContext, modelLocation));
+            var t4 = new Thread(() => TensorFlowUtils.GetModelSchema(mlContext, modelLocation));
+            var t5 = new Thread(() => TensorFlowUtils.GetModelSchema(mlContext, modelLocation));
+            t1.Start();
+            t2.Start();
+            t3.Start();
+            t4.Start();
+            t5.Start();
+            var a = model.GetInputSchema();
+            Assert.NotNull(a);
```
Comment (eerhardt) [inline test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs]: You may need to join all the threads you started, and then check to make sure they didn't throw any exceptions during the background thread.

**Comment by zeahmed** [comment]
It seems like OpType = "SaveV2" has the same problem.

**Comment by zeahmed** [comment]
Added a more generic test on op.NumOutputs.

---
In reply to: [455301754](https://github.com/dotnet/machinelearning/pull/2173#issuecomment-455301754) [](ancestors = 455301754)

File: src/Microsoft.ML.TensorFlow/TensorFlow/TensorflowUtils.cs [hunk 3]
```diff
@@ -40,7 +40,11 @@ internal static Schema GetModelSchema(IExceptionContext ectx, TFGraph graph, str
 
                 // If the type is not supported in ML.NET then we cannot represent it as a column in an Schema.
                 // We also cannot output it with a TensorFlowTransform, so we skip it.
-                if (mlType == null)
+                // Furthermore, 'NoOp' operator usually has 'Unknown' type.
+                // But sometimes it can have valid type upon querying model for multiple times such as 'Float'.
+                // This behavior is inconsistent in TensorFlow runtime (https://github.com/dotnet/machinelearning/issues/2156).
+                // Therefore, filtering all operators that have no outputs.
+                if (mlType == null || op.NumOutputs <= 0)
```
Comment (zeahmed) [inline src/Microsoft.ML.TensorFlow/TensorFlow/TensorflowUtils.cs]: >op.NumOutputs <= 0 [](start = 38, length = 18)

This is the more generic (as well as save to) test.

File: src/Microsoft.ML.TensorFlow/TensorFlow/TensorflowUtils.cs [hunk 4]
```diff
@@ -40,7 +40,11 @@ internal static Schema GetModelSchema(IExceptionContext ectx, TFGraph graph, str
 
                 // If the type is not supported in ML.NET then we cannot represent it as a column in an Schema.
                 // We also cannot output it with a TensorFlowTransform, so we skip it.
-                if (mlType == null)
+                // Furthermore, 'NoOp' operator usually has 'Unknown' type.
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.TensorFlow/TensorFlow/TensorflowUtils.cs]: >Furthermore, 'NoOp' operator usually has 'Unknown' type. [](start = 19, length = 56)

can you update this statement?

**Pull request merged**